File: Recursion/invert_binary_tree.py
#     Python
# Time complexity - O (N)
# Space complexity - O (H) - depth of recursion stack
# Worst (unbalanced) O(n) - , Best (balanced) - O(log(n))

class Solution:
    def invertTree(self, root: TreeNode) -> TreeNode:
        if root is None:
            return None

        root.left, root.right = root.right, root.left

        self.invertTree(root.left)
        self.invertTree(root.right)

        return root

File: Recursion/LeetCode - 226. Invert Binary Tree/InvertBinaryTree.java
//     Java
// Time complexity - O (N)
// Space complexity - O (H) - depth of recursion stack
// Worst (unbalanced) O(n) - , Best (balanced) - O(log(n))
public class InvertBinaryTree {
    public TreeNode invertTree(TreeNode root) {
        if (root == null) {
            return null;
        }

        TreeNode temp = root.left;
        root.left = root.right;
        root.right = temp;

        invertTree(root.left);
        invertTree(root.right);

        return root;
    }
}
